### DISCUSSION TITLE
Help with precedence

### DISCUSSION THREAD

**Discussion by wstevick** [tree-sitter/tree-sitter#2828] [Q&A] [answered]
I'm writing a parser for a functional language, where function application is looks like `a b`, and `a b c` is equivalent to `(a b) c`. I'm having an issue where `a - b - c` is interpreted as calling `a (-(b (-c)))` (i.e. calling `a` with the negative result of calling `b` with `-c`). I'd like to solve this using the precedence operators, but haven't been able to find a way.

**Comment by ahelwer** [comment]
What happens when you make your negative prefix operator have lower precedence than your infix minus operator?

**Comment by wstevick** [comment]
Doesn't help. Here're my precedences:
```js
precedences: () => [
  ["apply", "mult", "add", "neg", "trailing_exp"],
],
```

It's still parsing `x - y - z` as `(apply
          (name)
          (neg
            (apply
              (name)
              (neg
                (name)))))))`

**Comment by ahelwer** [comment]
So your apply operator is higher precedence than infix op expression operator? What about playing around with those?

**Comment by wstevick** [comment]
I can't figure out a way.

**Reply by ahelwer** [reply]
The way is to make your apply rule lower-precedence than your expression rule. If this isn't resolvable statically you might need to use dynamic precedence with GLR parsing by declaring a conflict.

**Reply by wstevick** [reply]
What do you mean? Are you saying to have the apply rule lower-precedence than the binary expression rule, or should I create a precedence level for expressions in general?

**Reply by ahelwer** [reply, marked as answer]
Yes, make your apply rule lower precedence than the binary expression rule and see what happens.

**Reply by wstevick** [reply]
I have `["neg", "mult", "add", "apply", "trailing_exp"]`. The syntax tree is `(apply (apply (name) (neg (name))) (neg (name)))`. Or, `(a (-b)) (-c)`.

Thank you so much for your help.
